test/unit/srv_option_handling.test.ts

```diff
@@ -0,0 +1,119 @@
+import { expect } from 'chai';
+import * as dns from 'dns';
+import * as sinon from 'sinon';
+import { promisify } from 'util';
+
+import { MongoCredentials } from '../../src/cmap/auth/mongo_credentials';
+import { $EXTERNAL_AUTH_SOURCE_MECHANISMS, AuthMechanism } from '../../src/cmap/auth/providers';
+import { resolveSRVRecord } from '../../src/connection_string';
+
+const resolveSRVRecordAsync = promisify(resolveSRVRecord);
+
+describe('Srv Option Handling', () => {
+  let resolveSrvStub: sinon.SinonStub;
+  let resolveTxtStub: sinon.SinonStub;
+  let resolveSRVRecordStub: sinon.SinonStub;
+  let lookupStub: sinon.SinonStub;
+
+  afterEach(async () => {
+    if (resolveSrvStub) {
+      resolveSrvStub.restore();
+      resolveSrvStub = undefined;
+    }
+    if (resolveTxtStub) {
+      resolveTxtStub.restore();
+      resolveTxtStub = undefined;
+    }
+    if (lookupStub) {
+      lookupStub.restore();
+      lookupStub = undefined;
+    }
+    if (resolveSRVRecordStub) {
+      resolveSRVRecordStub.restore();
+      resolveSRVRecordStub = undefined;
+    }
+  });
+
+  function makeStub(txtRecord: string) {
+    const mockAddress = [
+      {
+        name: 'localhost.test.mock.test.build.10gen.cc',
+        port: 2017,
+        weight: 0,
+        priority: 0
+      }
+    ];
+
+    const mockRecord: string[][] = [[txtRecord]];
+
+    // first call is for the driver initial connection
+    // second call will check the poller
+    resolveSrvStub = sinon.stub(dns, 'resolveSrv').callsFake((address, callback) => {
+      return process.nextTick(callback, null, mockAddress);
+    });
+
+    resolveTxtStub = sinon.stub(dns, 'resolveTxt').callsFake((address, thisIsWhatWeAreTesting) => {
+      thisIsWhatWeAreTesting(null, mockRecord);
+    });
+  }
+
+  for (const mechanism of $EXTERNAL_AUTH_SOURCE_MECHANISMS) {
+    it('should set authSource to $external for ${mechanism} external mechanism:', async function () {
+      makeStub('authSource=thisShouldNotBeAuthSource');
+      const options = {
+        credentials: new MongoCredentials({
+          source: '$external',
```

Comment:
since source is explicitly set as `$external` here, does it actually matter that a real external mechanism is used in the test? i.e., what would happen if we run this test with `source: '$external'` and `mechanism: 'potato'`? Also, what would happen if we use an external mechanism but set `source: 'admin'`?

Reply:
1) It matters since _if_ a non-external mechanism is used, an entirely new MongoCredentials instance would be called as found in `connection_string.ts`. So these cases check the `options.credentials` of a potentially different MongoCredentials instance not just whatever is in the options variable. Just added new assertions to test/show this! 

2) also above all we only want to test the SRV options handling case, not necessarily MongoCredentials in this scope